### PR TITLE
Ruby/Database capstone project

### PR DESCRIPTION
**Pull request by hienphan0111**
### Ruby/Database capstone project:

Our group consist of @hienphan0111 , @chepkok3 and @alicemirigo92 

[Click here](https://drive.google.com/file/d/17gfHgBK8T669CZzbS-xHRZo_-ayhyaaS/view?usp=share_link) to see the brief introduction of project video.

In this project, we will create a console app that will help user to keep a record of different types of things you own: books, music albums, movies, and games. Everything will be based on the UML class diagram presented below. The data will be stored in JSON files but you will also prepare a database with tables structure analogical to your program's class structure.

"Catalog of my things" should be a simple console app that allows users to manage collections of the things they own. It should be based on the following UML class diagram.
image

Logic
- Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- All methods visible in the diagram should be implemented:
 - add_item method (in all classes that have that method)
   - should take an instance of the Item class as an input.
   - should add the input item to the collection of items.
   - should add self as a property of the item object (by using the correct setter from the item object).
 - can_be_archived?() in the Item class
   - should return true if published_date is older than 10 years.
   - otherwise, it should return false.
 - move_to_archive() in the Item class
   - should reuse can_be_archived?() method.
   - should change the archived property to true if the result of the can_be_archived?() method is true.
   - should do nothing if the result of the can_be_archived?() method is false.
 - can_be_archived?() in the Book class
   - should override the method from the parent class.
   - should return true if parent's method returns true OR if cover_state equals to "bad".
   - otherwise, it should return false.
 - can_be_archived?() in the MusicAlbum class
   - should override the method from the parent class.
   - should return true if parent's method returns true AND if on_spotify equals true.
   - otherwise, it should return false.
 - can_be_archived?() in the Game class
   - should override the method from the parent class.
   - should return true if parent's method returns true AND if last_played_at is older than 2 years.
   - otherwise, it should return false.
- Add unit tests for all implemented methods.
User interface
 - Create a main.rb file that will serve as your console app entry-point.
 - Your console app, at the start, should:
    - Present the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
 - Have a way to quit the app.
    - The following options should be available:
    - List all books
    - List all music albums
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - Add a book
    - Add a music album
    - Add a game
- All data should be preserved by saving collections in .json files.
Database
  - Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot 
  - mplement inheritance in the database tables - add all properties and associations from the parent Item class as table 
  -  columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### CODE REVIEW

File: components/music_album.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    super(publish_date)
+    @on_spotify = on_spotify
+    @name = name
+    @genre = genre
+  end
+
+  def can_be_archived?
+    @on_spotify == true || super
+  end
+
```
Comment (ShadyShawkat) [inline components/music_album.rb]: - [ ] Kindly make sure that this `can_be_archived?` method returns `true` if both the parent's method "AND" the `on_spotify` equal `true`, as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=if%20parent%27s%20method%20returns%20true%20AND%20if%20on_spotify%20equals%20true.) requirement.

File: database/schema.sql [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	FOREIGN Key(label_id) REFERENCES label(id),
+    PRIMARY KEY(id)
+);
+
+create table genre(
+	id INT NOT NULL GENERATED ALWAYS AS IDENTITY,
+    name VARCHAR(100),
+    items TEXT[],
+    PRIMARY KEY(id)
+);
```
Comment (ShadyShawkat) [inline database/schema.sql]: - [ ] Great work on the Schema file. However, you forgot to create indexes for each of the created foreign keys. Kindly do so.

File: README.md [hunk 3]
```diff
@@ -27,8 +27,6 @@
```
Comment (ShadyShawkat) [inline README.md:27]: - [ ] Great work on the project. I can see you've put a lot of effort into it. Kindly rerecord the video presentation of the project to be no more than 5 minutes. The captured video is more than 10 minutes long which is too long and goes against [the following requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=maximum%20length%20of%205%20minutes).

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi Team, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- Good file structure✔️
- All tests are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 
- Your code is clean and well documented ✔️

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by franclobo — APPROVED** [review]
Hi @hienphan0111, @chepkok3, and @alicemirigo92 , 

# STATUS: APPROVED 😃 

git-merge

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- No linter errors ✔️ 
- Correct schema ✔️ 
- Correct Git flow ✔️ 
- Professional documentation ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**